zproject/backends.py

```diff
@@ -910,6 +938,71 @@ def sync_custom_profile_fields_from_ldap(
         except SyncUserError as e:
             raise ZulipLDAPError(str(e)) from e
 
+    def sync_groups_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUser) -> None:
+        if user_profile.realm.string_id not in settings.LDAP_GROUP_SYNC_LIST:
+            # no groups to sync for this realm
+            return
+
+        configured_ldap_group_names_for_sync = set(
+            settings.LDAP_GROUP_SYNC_LIST[user_profile.realm.string_id]
+        )
+
+        try:
+            ldap_logger.debug("Syncing groups for user: %s", user_profile.id)
+            intended_group_name_set_for_user = set(ldap_user.group_names).intersection(
+                configured_ldap_group_names_for_sync
+            )
+
+            existing_group_name_set_for_user = set(
+                UserGroupMembership.objects.filter(
+                    user_group__realm=user_profile.realm,
+                    user_group__name__in=set(
+                        settings.LDAP_GROUP_SYNC_LIST[user_profile.realm.string_id]
+                    ),
+                    user_profile=user_profile,
+                )
+                .distinct("user_group_id")
```

Comment:
Do we need this "distinct" statement here? We cannot have two `UserGroupMembership` objects with same user_profile  and user_group.

Reply:
not sure, it does look the same to me. I think we can remove it.

Reply:
https://github.com/zulip/zulip/pull/23039/commits/8e700f07d83c91bd32441afe1af1c9212466a4ff